Host helper C++ class

Trinitou opened the pull request:
Adding a C++ helper class CLAP host developers can easily derive from

Left out here for future PRs / discussion
- misbehaviour checks, e.g. some ensureMainThread / ensureAudioThread methods, checking level templating etc.
- clap-helpers repository description: "Minimal set of helpers to implement your plugin" would need to change to something more welcoming also for host devs -> or should this be split into two separate repos instead ("clap-plugin-helpers" and "clap-host-helpers")?
- documentation
- test

Trinitou commented:
Probably I'll focus on the `clap::helpers::Host` first and leave out the PluginProxy stuff (can go into a separate PR). Main goal would be to extract the reusable stuff from `PluginHost` in https://github.com/free-audio/clap-host/blob/main/host/plugin-host.hh so that `PluginHost` can be derived from the new `clap::helpers::Host` class

abique commented:
Good luck with this task, this isn't an easy one. The host'll be harder to generalize/glue than the plugin.

Trinitou commented:
- [x] reverted the PluginProxy changes
- [x] removed the template code (at least for now)
- [x] added all extensions necessary to derive the PluginHost class from the `free-audio/plugin-host` repo

(Edit: moved open todos into first comment)

Please also have a look into this commit in my fork of the `plugin-host` repo:
[f22ed540aa206c2452bdfb9011a76cf3b14cd9f2](https://github.com/Trinitou/clap-host/commit/f22ed540aa206c2452bdfb9011a76cf3b14cd9f2)
This is how I think it could look like using the new `clap::helpers::Host` class. Maybe some details are still missing. What do you think @abique ?

Trinitou commented:
I think we could leave it like this for now except you want to have misbehavior checks from the beginning.

abique commented:
> Please also have a look into this commit in my fork of the `plugin-host` repo:
> [f22ed540aa206c2452bdfb9011a76cf3b14cd9f2](https://github.com/Trinitou/clap-host/commit/f22ed540aa206c2452bdfb9011a76cf3b14cd9f2)
> This is how I think it could look like using the new `clap::helpers::Host` class. Maybe some details are still missing. What do you think @abique ?

That'd work for me. Thank you for doing this!

abique commented:
> I think we could leave it like this for now except you want to have misbehavior checks from the beginning.

I can see that you still have some checkboxes opened. Whenever you think it is ready to be merged I'll give it a review and merge it.

Trinitou commented:
Ok, @abique I looked through it and think it's ready enough for a review.

Cleaned up the checkboxes / comments. In the first comment I collected some ideas for future work / PRs.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/free-audio/clap-helpers?pullRequest=24) <br/>All committers have signed the CLA.

Trinitou commented:
One thing: In commit 5180427482efd8fc71ccb15e929b14c9ce498dc5 inside the `::from` method, I replaced the `std::cerr` + `std::terminate` with `throw` which I copied from the `clap-host` repo. Since the from method now is `noxcept`, this will cause a warning I think. Should we better switch back to std::terminate?

abique commented:
@Trinitou Thank you for this work.
I'll review it in the afternoon.

abique commented on include/clap/helpers/host.hh:
```diff
@@ -0,0 +1,176 @@
+      static void tailChanged() noexcept {}
+
+      // clap_host_thread_check
+      virtual bool implementsThreadCheck() const noexcept { return false; }
```
I'd make the thread check implementation mandatory maybe? :thinking:

abique replied:
It isn't wrong to have it optional, but you loose 95% of the checks if you don't have the thread check. I think it isn't reasonable to start implementing a host without it.
Please make `threadCheckIsMainThread()` and `threadCheckIsAudioThread()` pure virtual and get rid of `implementsThreadCheck()`.

abique commented on include/clap/helpers/host.hh:
```diff
@@ -0,0 +1,176 @@
+      /////////////////////////
+
+      // clap_host
+      virtual void requestRestart() noexcept {}
```
Those 3 host methods aren't optional, so they should be pure virtual.

abique commented on include/clap/helpers/host.hh:
```diff
@@ -0,0 +1,176 @@
+
+      // clap_host_log
+      virtual bool implementsLog() const noexcept { return false; }
+      virtual void logLog(clap_log_severity severity, const char *message) const noexcept {}
```
default implementation here should write to `std::clog` maybe?

abique replied:
Maybe `clog` isn't the right output, but something should go on the console at least by default.

abique replied:
Ah actually I'm not sure, given that `implementsLog()` returns false by default.
The current code is fine :ok_hand:
